[JSC] Add fast path to TypedArray.from

Constellation opened the pull request:
#### 1b440efcb4ae8dbc2a031e14ad222cc411e6e61b
<pre>
[JSC] Add fast path to TypedArray.from
<a href="https://bugs.webkit.org/show_bug.cgi?id=239936">https://bugs.webkit.org/show_bug.cgi?id=239936</a>

Reviewed by Mark Lam, Ross Kirsling and Keith Miller.

This patch adds a fast path to TypedArray.from. Similar to what we do with Array, we integrate TypedArray&apos;s iterator-protocol WatchpointSet.
And now TypedArray has isIteratorProtocolFastAndNonObservable() function which answers whether we can do fast-iteration without
user observable behavior withhout using normal generic for-of iterator protocol.

The most complicated part of this WatchpointSet is that &quot;length&quot; property of TypedArray. Unlike Array, TypedArray does not have
its own &quot;length&quot; property. Instead it has &quot;length&quot; getter in TypedArray.prototype. Thus, we need to ensure the following.

1. Uint8Array instance does not have its own &quot;length&quot; property
2. Uint8Array instance&apos;s [[Prototype]] is Uint8Array.prototype
3. Uint8Array.prototype does not have &quot;length&quot; property
4. Uint8Array.prototype&apos;s [[Prototype]] is TypedArray.prototype
5. TypedArray.prototype has default &quot;length&quot; getter

We set &quot;length&quot; absence ObjectPropertyCondition (OPC) on Uint8Array.prototype. And we also set &quot;length&quot; equivalence OPC on TypedArray.prototype.length.
As the same with Array&apos;s iterator protocol, we also ensure @@iterator function&apos;s absence status on Uint8Array.prototype and TypedArray.prototype too.
By installing these watchpoints, we can quickly answer whether this TypedArray instance can get fast iteration.

And in TypedArray.from, we use this condition to bypass normal for-of iteration protocol. Also we extend this function to accept normal Array: using
Array&apos;s iterator protocol WatchpointSet to make it fast for Int32 and Double Arrays.

TypedArray.from from TypedArray gets 396.0x faster. And TypedArray.from from Array gets 6.8x faster.

                                   ToT                     Patched

typed-array-from-array      913.8926+-1.7045     ^    133.6719+-0.1703        ^ definitely 6.8368x faster
typed-array-from           1183.6522+-2.1020     ^      2.9890+-0.3176        ^ definitely 396.0042x faster

* microbenchmarks/typed-array-from.js: Added.
* stress/typed-array-from-array-iterator-protocol.js: Added.
(shouldBe):
(shouldBeArray):
(values.__proto__.next):
* stress/typed-array-from.js: Added.
(shouldBe):
(shouldBeArray):
(shouldThrow):
(shouldBeArray.TestArray):
(Uint8Array.prototype.__proto__.Symbol.iterator):
* Source/JavaScriptCore/builtins/BuiltinNames.h:
* Source/JavaScriptCore/builtins/TypedArrayConstructor.js:
* Source/JavaScriptCore/bytecode/LinkTimeConstant.h:
* Source/JavaScriptCore/runtime/JSGlobalObject.cpp:
(JSC::JSGlobalObject::init):
* Source/JavaScriptCore/runtime/JSTypedArrayViewPrototype.cpp:
(JSC::isTypedArrayViewConstructor):
(JSC::JSC_DEFINE_HOST_FUNCTION):
(JSC::JSTypedArrayViewPrototype::finishCreation):
* Source/JavaScriptCore/runtime/JSTypedArrayViewPrototype.h:

Canonical link: <a href="https://commits.webkit.org/252976@main">https://commits.webkit.org/252976@main</a>
</pre>

kmiller68 commented on Source/JavaScriptCore/ChangeLog:
```diff
@@ -1,3 +1,30 @@
+        Reviewed by NOBODY (OOPS!).
+
+        This patch adds a fast path to TypedArray.from. When the argument is TypedArray and @@iterator is TypedArray#values,
+        then it is ensured that we can iterate this typed array without side effect (except for detached case. But it is handled
```
Nit: But it is handled => But this is handled

Constellation replied:
Done.

kmiller68 commented on Source/JavaScriptCore/ChangeLog:
```diff
@@ -1,3 +1,30 @@
+        then it is ensured that we can iterate this typed array without side effect (except for detached case. But it is handled
+        in the fast path too). Thus, we can add a fast path using TypedArray#set to bypass iteration based on iterator protocol.
+
+        This change improves TypedArray.from faster by 134x.
```
Nit: delete faster or /improves/makes/s

Constellation replied:
Done.

kmiller68 reviewed (approved):
r=me with some typos

Constellation commented on Source/JavaScriptCore/builtins/TypedArrayConstructor.js:
```diff
@@ -64,6 +65,17 @@ function from(items /* [ , mapfn [ , thisArg ] ] */)
+        // since TypedArray does not accept any accessor definitions for indexed properties.
+        // We also need to ensure that |this| constructor will not cause user observable side-effect. Otherwise, it is possible
+        // that |this| constructor detaches arrayLike before using.
+        if (!mapFn && @isTypedArrayView(arrayLike) && iteratorMethod === @typedArrayValues && @isTypedArrayViewConstructor(this)) {
```
Oops, I think I need a bit more checks here. Let's consider that we override iterator's prototype.

```
%ArrayIterator%.prototype.next = function () { return { value: 42, done: false }; };
```
Then, we should use this function to iterate typed array to collect values.
This means that we need to ensure that array-iterator-protocol is not broken here. `isArrayPrototypeIteratorProtocolFastAndNonObservable` should be checked.
I'll extend this patch to add this and test (After that, I'll request another review).

Constellation replied:
I've added `canUseFastTypedArrayFrom` function.

Constellation replied:
Ugh, we still need to validate a bit more: we should check whether this `arrayLike` does not have `length` / index own property. Fix forthcoming.

Constellation replied:
We need TypedArray iterator protocol validation. Because,

1. We can still put `length` value / getter in own typed array or `%TypedArray%.prototype`.
2. On the other hand, indexed access cannot be trapped & overridden

We should have sane chain watchpoint for typed array prototypes. And we should also have a watchpoint whether `length` gets overridden in prototypes.
Also, we need to check that we do not have `length` own property on `arrayLike`.

rkirsling commented on Source/JavaScriptCore/runtime/JSGlobalObject.cpp:
```diff
@@ -1441,11 +1441,14 @@ capitalName ## Constructor* lowerName ## Constructor = featureFlag ? capitalName
+            init.set(JSFunction::create(init.vm, jsCast<JSGlobalObject*>(init.owner), 1, String(), typedArrayViewPrivateFuncIsTypedArrayView, IsTypedArrayViewIntrinsic));
         });
     m_linkTimeConstants[static_cast<unsigned>(LinkTimeConstant::isSharedTypedArrayView)].initLater([] (const Initializer<JSCell>& init) {
             init.set(JSFunction::create(init.vm, jsCast<JSGlobalObject*>(init.owner), 0, String(), typedArrayViewPrivateFuncIsSharedTypedArrayView));
```
Seems like this one should be changed too?
```suggestion
            init.set(JSFunction::create(init.vm, jsCast<JSGlobalObject*>(init.owner), 1, String(), typedArrayViewPrivateFuncIsSharedTypedArrayView));
```

Constellation replied:
Fixed.

rkirsling reviewed (approved):
This is incredible! Patch looks good from my perspective, but I should probably not be the only reviewer since I'm still not super used to watchpoint-related code.

MenloDorian commented on Source/JavaScriptCore/runtime/JSGlobalObjectInlines.h:
```diff
@@ -90,6 +90,31 @@ ALWAYS_INLINE bool JSGlobalObject::isArrayPrototypeIteratorProtocolFastAndNonObs
+
+    typedArrayPrototype(typedArrayType); // Materialize WatchpointSet.
+
+    // Since TypedArray iterator is the same to ArrayIterator, we need to check the state of ArrayIteratorProtocolWatchpointSet.
```
/TypedArray iterator is the same to ArrayIterator/TypedArray iteration uses ArrayIterator/

Constellation replied:
Fixed.

MenloDorian commented on Source/JavaScriptCore/runtime/JSGlobalObjectInlines.h:
```diff
@@ -90,6 +90,31 @@ ALWAYS_INLINE bool JSGlobalObject::isArrayPrototypeIteratorProtocolFastAndNonObs
+    typedArrayPrototype(typedArrayType); // Materialize WatchpointSet.
+
+    // Since TypedArray iterator is the same to ArrayIterator, we need to check the state of ArrayIteratorProtocolWatchpointSet.
+    // But we do not need ot check isHavingABadTime() and array prototype's chain.
```
/ot/to/

Constellation replied:
Fixed.

MenloDorian commented on Source/JavaScriptCore/runtime/JSGlobalObjectInlines.h:
```diff
@@ -90,6 +90,31 @@ ALWAYS_INLINE bool JSGlobalObject::isArrayPrototypeIteratorProtocolFastAndNonObs
+    if (typedArrayIteratorProtocolWatchpointSet(typedArrayType).state() != IsWatched)
+        return false;
+
+    // This WatchpointSet ensures TypedArray.prototype has default "length" getter and @@iterator function.
```
/ensures/ensures that/

Constellation replied:
Fixed.

MenloDorian commented on Source/JavaScriptCore/runtime/JSGlobalObjectInlines.h:
```diff
@@ -90,6 +90,31 @@ ALWAYS_INLINE bool JSGlobalObject::isArrayPrototypeIteratorProtocolFastAndNonObs
+    if (!arrayIteratorProtocolWatchpointSet().isStillValid())
+        return false;
+
+    // This WatchpointSet ensures
```
/ensures/ensures that/

Constellation replied:
Fixed.

MenloDorian commented on Source/JavaScriptCore/runtime/JSGlobalObject.cpp:
```diff
@@ -2596,6 +2599,58 @@ void JSGlobalObject::installTypedArrayConstructorSpeciesWatchpoint(JSTypedArrayV
+{
+    VM& vm = this->vm();
+
+    DeferTermination deferScope(vm);
```
nit: let's use DeferTerminationForAWhile instead.   It's probably a good idea to always use DeferTerminationForAWhile instead of DeferTermination (which can throw the TerminationException on destruction).  Maybe we should get rid of DeferTermination as I can't think of any reason why we would really need it.

Constellation replied:
Right, changed.

MenloDorian commented on Source/JavaScriptCore/runtime/JSGlobalObject.cpp:
```diff
@@ -2596,6 +2599,58 @@ void JSGlobalObject::installTypedArrayConstructorSpeciesWatchpoint(JSTypedArrayV
+    ObjectPropertyCondition iteratorCondition = absenceCondition(vm.propertyNames->iteratorSymbol);
+
+    if (!lengthCondition.isWatchable(PropertyCondition::EnsureWatchability) || !iteratorCondition.isWatchable(PropertyCondition::EnsureWatchability)) {
+        typedArrayIteratorProtocolWatchpointSet(typedArrayType).invalidate(vm, StringFireDetail("Was not able to set up length watchpoint."));
```
Regarding "Was not able to set up length watchpoint.", is this accurate?  For one, we're checking both lengthCondition and iteratorCondition.  Would it be more accurate to say something like "Length or @@iterator absence is not watchable."?

Constellation replied:
This should be `iterator protocol watchpoint`.

MenloDorian commented on Source/JavaScriptCore/runtime/JSGlobalObject.cpp:
```diff
@@ -2596,6 +2599,58 @@ void JSGlobalObject::installTypedArrayConstructorSpeciesWatchpoint(JSTypedArrayV
+    }
+
+    RELEASE_ASSERT(!typedArrayIteratorProtocolWatchpointSet(typedArrayType).isBeingWatched());
+    typedArrayIteratorProtocolWatchpointSet(typedArrayType).touch(vm, "Set up length watchpoint.");
```
Is this "Set up length watchpoint." message accurate?

Constellation replied:
This should be a bit more generic `iterator protocol watchpoint`.

MenloDorian commented on Source/JavaScriptCore/runtime/JSGlobalObject.cpp:
```diff
@@ -2596,6 +2599,58 @@ void JSGlobalObject::installTypedArrayConstructorSpeciesWatchpoint(JSTypedArrayV
+    RELEASE_ASSERT(!typedArrayIteratorProtocolWatchpointSet(typedArrayType).isBeingWatched());
+    typedArrayIteratorProtocolWatchpointSet(typedArrayType).touch(vm, "Set up length watchpoint.");
+
+    typedArrayPrototypeLengthAbsenceWatchpoint(typedArrayType) = makeUnique<ObjectAdaptiveStructureWatchpoint>(this, lengthCondition, typedArrayIteratorProtocolWatchpointSet(typedArrayType));
+    typedArrayPrototypeLengthAbsenceWatchpoint(typedArrayType)->install(vm);
+    typedArrayPrototypeSymbolIteratorAbsenceWatchpoint(typedArrayType) = makeUnique<ObjectAdaptiveStructureWatchpoint>(this, iteratorCondition, typedArrayIteratorProtocolWatchpointSet(typedArrayType));
+    typedArrayPrototypeSymbolIteratorAbsenceWatchpoint(typedArrayType)->install(vm);
```
Where is the part that makes `typedArrayIteratorProtocolWatchpointSet` ensure this:
```
3. derived TypedArray prototype's [[Prototype]] is TypedArray.prototype.
```

Constellation replied:
ObjectPropertyCondition absence also ensures `[[Prototype]]`.

MenloDorian replied:
Got it.  Thanks.

MenloDorian commented on Source/JavaScriptCore/runtime/JSGenericTypedArrayViewPrototypeFunctions.h:
```diff
@@ -656,6 +658,69 @@ ALWAYS_INLINE EncodedJSValue genericTypedArrayViewPrivateFuncSort(VM& vm, JSGlob
+                if (LIKELY(!!value))
+                    result->setIndexQuicklyToNativeValue(i, ViewClass::Adaptor::toNativeFromInt32(value.asInt32()));
+                else
+                    result->setIndexQuickly(i, jsUndefined());
```
Is this valid? From what I can tell, `setIndexQuickly` can only handle a JSValue that is int32 or double.  See `Adaptor::Type toNativeFromValue(JSValue value)`.

Constellation replied:
Oops, it should be `toNativeFromValue(globalObject, ...` conversion.

MenloDorian replied:
We already know the source came from an Int32Shape contiguous array, and in this case, we have a hole.  Why go through a function to convert `undefined` to a value?  Isn't it always 0?

Constellation replied:
If the array is Int32Shape and iterator protocol is ensured, then we are ensured that we will get an undefined value here.
But what value will be stored into `result` depends on `ViewClass`. For example, if it is `Int32Array`, `0` will be stored. But if it is `Float64Array`, then it should get `NaN` since undefined will be converted to `NaN` when converting it to number.
So, that's why we need `toNativeFromValue`. It is ensured that this behavior is unobservable since it will not invoke any functions. Converting `undefined` to `ViewClass`'s type. But still only way we have is `toNativeFromValue`.
I wonder if we should add `toNativeFromUndefined`, which would be useful in the other place too. Trying.

Constellation replied:
Ah, wait, it is fun. BigInt64Array can get an error if `undefined` is converted! So, I think we should skip this optimization path for BigInt64Array etc. But it is fine since it cannot accept these Int32Shape etc. anyway

Constellation replied:
I'll fix it soon.

Constellation replied:
Ah, and it just works with `toNativeFromDouble(PNaN)`.

MenloDorian commented on Source/JavaScriptCore/runtime/JSGenericTypedArrayViewPrototypeFunctions.h:
```diff
@@ -656,6 +658,69 @@ ALWAYS_INLINE EncodedJSValue genericTypedArrayViewPrivateFuncSort(VM& vm, JSGlob
+                if (LIKELY(!std::isnan(d)))
+                    result->setIndexQuicklyToNativeValue(i, ViewClass::Adaptor::toNativeFromDouble(d));
+                else
+                    result->setIndexQuickly(i, jsUndefined());
```
Ditto.

MenloDorian replied:
Ditto here: we know that the source came from a DoubleShape contiguous array.  So, a hole should convert to some constant (though I'm not sure what the spec says it is), right?

MenloDorian commented on Source/JavaScriptCore/runtime/JSGenericTypedArrayView.h:
```diff
@@ -81,6 +81,7 @@ JS_EXPORT_PRIVATE const ClassInfo* getBigUint64ArrayClassInfo();
 //     static int8_t toNativeFromInt32(int32_t);
 //     static int8_t toNativeFromUint32(uint32_t);
 //     static int8_t toNativeFromDouble(double);
+//     static int8_t toNativeFromUndefined();
```
This appears to be unimplemented and unused.  I think you should implement it and use it.  See below for details.

Constellation replied:
changed.

MenloDorian commented on Source/JavaScriptCore/runtime/JSGenericTypedArrayViewPrototypeFunctions.h:
```diff
@@ -669,6 +669,70 @@ ALWAYS_INLINE EncodedJSValue genericTypedArrayViewPrivateFuncSort(VM& vm, JSGlob
+                if (LIKELY(!!value))
+                    result->setIndexQuicklyToNativeValue(i, ViewClass::Adaptor::toNativeFromInt32(value.asInt32()));
+                else
+                    result->setIndexQuicklyToNativeValue(i, ViewClass::Adaptor::toNativeFromDouble(PNaN));
```
For integer type arrays, the hole maps to `undefined`, which in turn maps to 0.  Going through `toNativeFromDouble` incurs a non-zero cost as it's not constexpr, and when `HAVE(FJCVTZS_INSTRUCTION)`, it routes through inline asm.  Let's implement `toNativeFromUndefined` as constexpr and have it return a 0 for integer type arrays, and PNaN for float type arrays, and use `toNativeFromUndefined` instead.

Constellation replied:
OK, added.

MenloDorian reviewed (approved):
r=me with improvements.

Constellation committed:
[JSC] Add fast path to TypedArray.from
https://bugs.webkit.org/show_bug.cgi?id=239936

Reviewed by Mark Lam, Ross Kirsling and Keith Miller.

This patch adds a fast path to TypedArray.from. Similar to what we do with Array, we integrate TypedArray's iterator-protocol WatchpointSet.
And now TypedArray has isIteratorProtocolFastAndNonObservable() function which answers whether we can do fast-iteration without
user observable behavior withhout using normal generic for-of iterator protocol.

The most complicated part of this WatchpointSet is that "length" property of TypedArray. Unlike Array, TypedArray does not have
its own "length" property. Instead it has "length" getter in TypedArray.prototype. Thus, we need to ensure the following.

1. Uint8Array instance does not have its own "length" property
2. Uint8Array instance's [[Prototype]] is Uint8Array.prototype
3. Uint8Array.prototype does not have "length" property
4. Uint8Array.prototype's [[Prototype]] is TypedArray.prototype
5. TypedArray.prototype has default "length" getter

We set "length" absence ObjectPropertyCondition (OPC) on Uint8Array.prototype. And we also set "length" equivalence OPC on TypedArray.prototype.length.
As the same with Array's iterator protocol, we also ensure @@iterator function's absence status on Uint8Array.prototype and TypedArray.prototype too.
By installing these watchpoints, we can quickly answer whether this TypedArray instance can get fast iteration.

And in TypedArray.from, we use this condition to bypass normal for-of iteration protocol. Also we extend this function to accept normal Array: using
Array's iterator protocol WatchpointSet to make it fast for Int32 and Double Arrays.

TypedArray.from from TypedArray gets 396.0x faster. And TypedArray.from from Array gets 6.8x faster.

                                   ToT                     Patched

typed-array-from-array      913.8926+-1.7045     ^    133.6719+-0.1703        ^ definitely 6.8368x faster
typed-array-from           1183.6522+-2.1020     ^      2.9890+-0.3176        ^ definitely 396.0042x faster

* microbenchmarks/typed-array-from.js: Added.
* stress/typed-array-from-array-iterator-protocol.js: Added.
(shouldBe):
(shouldBeArray):
(values.__proto__.next):
* stress/typed-array-from.js: Added.
(shouldBe):
(shouldBeArray):
(shouldThrow):
(shouldBeArray.TestArray):
(Uint8Array.prototype.__proto__.Symbol.iterator):
* Source/JavaScriptCore/builtins/BuiltinNames.h:
* Source/JavaScriptCore/builtins/TypedArrayConstructor.js:
* Source/JavaScriptCore/bytecode/LinkTimeConstant.h:
* Source/JavaScriptCore/runtime/JSGlobalObject.cpp:
(JSC::JSGlobalObject::init):
* Source/JavaScriptCore/runtime/JSTypedArrayViewPrototype.cpp:
(JSC::isTypedArrayViewConstructor):
(JSC::JSC_DEFINE_HOST_FUNCTION):
(JSC::JSTypedArrayViewPrototype::finishCreation):
* Source/JavaScriptCore/runtime/JSTypedArrayViewPrototype.h:

Canonical link: https://commits.webkit.org/252976@main
